### PR TITLE
useragentprocessor to enrich user-agent data

### PR DESCRIPTION
**Pull request by nagarajatantry**
**Description:** <Describe what has changed.>
Adding a new processor `useragentprocessor`. It reads the userAgent tag from spans, extracts browser/os/device information and adds them as attributes

**Link to tracking Issue:** <Issue number if applicable>
https://github.com/open-telemetry/opentelemetry-collector/issues/1606

**Testing:** <Describe what testing was performed and which tests were added.>
I have added unit tests with > 85% coverage. Built the docker image and tested it locally.
```
make docker-otelcontribcol
docker run --rm -p 8080:8080 -p 8443:8443 -p 55678:55678 -p 55679:55679 -p 8888:8888 \
        -v "${PWD}/local/config.yaml":/otel-local-config.yaml \
       -v "${PWD}/processor/useragentprocessor/resources/regexes.yaml":/useragent-regexes.yaml \
       --name otelcol otelcontribcol:latest \
        --config otel-local-config.yaml \
        --metrics-addr=0.0.0.0:8888
```

**Documentation:** <Describe the documentation added.>
Documentation is added [here](https://github.com/tannaga/opentelemetry-collector-contrib/blob/master/processor/useragentprocessor/README.md).

Please review and provide feedback. Thank you for your time!

### CODE REVIEW

File: processor/useragentprocessor/useragentprocessor.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+func (u *userAgentProcessor) enrichSpanWithUserAgent(userAgent string, span pdata.Span) {
+	c := u.uaParser.Parse(userAgent)
+	span.Attributes().Insert("browserFamily", pdata.NewAttributeValueString(c.UserAgent.Family))
+	span.Attributes().Insert("browserVersion", pdata.NewAttributeValueString(c.UserAgent.ToVersionString()))
+	span.Attributes().Insert("deviceFamily", pdata.NewAttributeValueString(c.Device.Family))
+	span.Attributes().Insert("deviceBrand", pdata.NewAttributeValueString(c.Device.Brand))
+	span.Attributes().Insert("deviceModel", pdata.NewAttributeValueString(c.Device.Model))
+	span.Attributes().Insert("osFamily", pdata.NewAttributeValueString(c.Os.Family))
+	span.Attributes().Insert("osVersion", pdata.NewAttributeValueString(c.Os.ToVersionString()))
+}
```
Comment (bogdandrutu) [inline processor/useragentprocessor/useragentprocessor.go:70]: Some of the conventions are defined here:
https://github.com/open-telemetry/opentelemetry-collector/blob/master/translator/conventions/opentelemetry.go#L19

Some are missing in the specification (probably need to add them):
https://github.com/open-telemetry/opentelemetry-specification/tree/master/specification/resource/semantic_conventions#todos

Comment (bogdandrutu) [reply processor/useragentprocessor/useragentprocessor.go:70]: Also they should be part of the "Resource" not Span :)

Comment (tannaga) [reply processor/useragentprocessor/useragentprocessor.go]: @bogdandrutu thank you for the feedback. Moved the logic to look for usrAgent in `Resource` and added the new attributes in that section.
Do you recommend me to propose the new tags under `conventions` first in a separate PR?

Comment (bogdandrutu) [reply processor/useragentprocessor/useragentprocessor.go]: To make progress, I think we should reduce the scope of this PR to only elements that are defined in the specs, then once we have specs for others we add them. What do you think? This way I think we don't block progress and we allow future improvements

Comment (tannaga) [reply processor/useragentprocessor/useragentprocessor.go]: Sounds good. Under conventions, i currently see couple of them `os.type` and `os.description`. 
https://github.com/open-telemetry/opentelemetry-collector/blob/master/translator/conventions/opentelemetry.go#L56-L57

Could you please recommend the right forum where we can discuss about adding new fields to the specs?

**Comment by tigrannajaryan** [comment]
@tannaga please sign the CLA and make sure EasyCLA passes.

**Comment by tannaga** [comment]
> @tannaga please sign the CLA and make sure EasyCLA passes.

Couple of my commits were made using wrong email address. I am trying to use rebase and amend author options, but I am running into all sort of conflicts in `go.sum`.  Do you know if there is an easy way to fix the email address for the 2 commits?

File: processor/useragentprocessor/useragentprocessor.go (hunk 1 above)
Comment (tannaga) [reply processor/useragentprocessor/useragentprocessor.go]: > > @tannaga please sign the CLA and make sure EasyCLA passes.
> 
> Couple of my commits were made using wrong email address. I am trying to use rebase and amend author options, but I am running into all sort of conflicts in `go.sum`. Do you know if there is an easy way to fix the email address for the 2 commits?

I tried few things to fix the author of my earlier conflicts, but it seems a little messy. I will close this PR and submit a new one.
cc @bogdandrutu @tigrannajaryan

**Comment by bogdandrutu** [comment]
@tannaga you neeed to re-create the commits, or squash them. We cannot merge as long as the CLA test is not passing.

**Comment by tannaga** [comment]
Will create a new PR  to resolve the EasyCLA test.

**Comment by SaschaBrechmannVHV** [comment]
Hi @tannaga , how about this PR, i like the function and would love to see it in Upstream.
Regards, Sascha

**Comment by faustodavid** [comment]
Hey @tannaga this functionality looks awsome. Will you reopen this PR?

**Pull request closed without merging**